src/python/pants/backend/python/dependency_inference/default_module_mapping.py

```diff
@@ -159,7 +189,18 @@ def two_groups_hyphens_two_replacements_with_suffix(
     "websocket-client": ("websocket",),
 }
 
-DEFAULT_TYPE_STUB_MODULE_MAPPING = {
+DEFAULT_TYPE_STUB_MODULE_PATTERN_MAPPING: Dict[re.Pattern, List[Callable[[Match[str]], str]]] = {
+    re.compile(r"""^stubs_(.+)"""): [first_group_hyphen_to_underscore],
+    re.compile(r"""^types_(.+)"""): [first_group_hyphen_to_underscore],
+    re.compile(r"""^stubs-(.+)"""): [first_group_hyphen_to_underscore],
+    re.compile(r"""^types-(.+)"""): [first_group_hyphen_to_underscore],
+    re.compile(r"""^(.+)_stubs"""): [first_group_hyphen_to_underscore],
+    re.compile(r"""^(.+)_types"""): [first_group_hyphen_to_underscore],
+    re.compile(r"""^(.+)-stubs"""): [first_group_hyphen_to_underscore],
+    re.compile(r"""^(.+)-types"""): [first_group_hyphen_to_underscore],
```

Comment:
these do the work of the `fallback[6:]`/`fallback[:6]` before, without needing to check for those/assuming prefix length